### PR TITLE
Room : Add an external users management for rooms

### LINKED ISSUE #247: Room : Add an external users management for rooms
Related to : 
- dinsic-pim/tchap-android#474
- dinsic-pim/tchap-android#476
- dinsic-pim/tchap-android#484
- dinsic-pim/tchap-android#475
- dinsic-pim/tchap-android#483
- matrix-org/matrix-dinsic#288

The linked PR will implement all these features.

### PR DESCRIPTION
**Pull request by Yoshiin**
Resolve #247 

- [x] dinsic-pim/tchap-android#474
- [x] dinsic-pim/tchap-android#476
- [x] dinsic-pim/tchap-android#484
- [x] dinsic-pim/tchap-android#483
- [x] dinsic-pim/tchap-android#475

More PRs will be added. 
Do not merge until all features are commited.

### CODE REVIEW

**Comment by Yoshiin** [comment]
All done. Can now be reviewed.

**Comment by Yoshiin** [comment]
All done. Can now be reviewed/merged.

File: src/components/views/dialogs/AddressPickerDialog.js [hunk 1]
```diff
@@ -367,6 +368,19 @@ module.exports = React.createClass({
...
     },
 
+    _getAccessRules: function(roomId) {
+        const stateEventType = "im.vector.room.access_rules";
+        const keyName = "rule";
+        const defaultValue = "restricted";
+        const room = MatrixClientPeg.get().getRoom(roomId);
+        const event = room.currentState.getStateEvents(stateEventType, '');
+        if (!event) {
+            return defaultValue;
```
[the remark this answers is not in the document]
Comment (giomfo) [reply src/components/views/dialogs/AddressPickerDialog.js]: Maybe this method `_getAccessRules` should be implemented only once somewhere, by considering all the possibilities:
if (access_rule is available)
   return access_rule
else if (room.isDirect)
   return `direct`
else
   return `restricted`

File: src/components/views/dialogs/AddressPickerDialog.js [hunk 2]
```diff
@@ -426,14 +447,23 @@ module.exports = React.createClass({
...
-            suggestedList.unshift({
-                addressType: addrType,
-                address: query,
-                isKnown: false,
-            });
             if (this._cancelThreepidLookup) this._cancelThreepidLookup();
             if (addrType === 'email') {
-                this._lookupThreepid(addrType, query).done();
+                this._lookupThreepid(addrType, query).then(val => {
+                    if (val !== false) {
```
Comment (giomfo) [inline src/components/views/dialogs/AddressPickerDialog.js]: Don't you think the correct check here should be if (val == null)?

File: src/components/views/dialogs/AddressPickerDialog.js [hunk 3]
```diff
@@ -426,14 +447,23 @@ module.exports = React.createClass({
...
+                            addressType: addrType,
+                            address: query,
+                            isKnown: false,
+                        });
+                    }
+                    self.setState({
+                        suggestedList,
+                        error: false,
+                    }, () => {
+                        if (self.addressSelector) self.addressSelector.moveSelectionTop();
```
Comment (giomfo) [inline src/components/views/dialogs/AddressPickerDialog.js]: Are you sure this moveSelectionTop  is required here

Comment (Yoshiin) [reply src/components/views/dialogs/AddressPickerDialog.js]: Is not.

File: src/components/views/dialogs/AddressPickerDialog.js [hunk 4]
```diff
@@ -392,8 +393,15 @@ module.exports = React.createClass({
     },
 
     _processResults: function(results, query) {
+        const self = this;
```
Comment (giomfo) [inline src/components/views/dialogs/AddressPickerDialog.js]: Why do you define `self` here?

File: src/components/views/dialogs/AddressPickerDialog.js [hunk 5]
```diff
@@ -426,14 +434,23 @@ module.exports = React.createClass({
...
-            suggestedList.unshift({
-                addressType: addrType,
-                address: query,
-                isKnown: false,
-            });
             if (this._cancelThreepidLookup) this._cancelThreepidLookup();
             if (addrType === 'email') {
-                this._lookupThreepid(addrType, query).done();
+                this._lookupThreepid(addrType, query).then(val => {
+                    if (val !== false) {
```
Comment (giomfo) [inline src/components/views/dialogs/AddressPickerDialog.js]: I renew my comment here "Don't you think the correct check here should be if (val == null)?"

File: src/components/views/dialogs/AddressPickerDialog.js [hunk 6]
```diff
@@ -426,14 +434,23 @@ module.exports = React.createClass({
...
+                    self.setState({
+                        suggestedList,
+                        error: false,
+                    }, () => {
+                        if (self.addressSelector) self.addressSelector.moveSelectionTop();
+                    });
+                });
             }
         }
         this.setState({
```
Comment (giomfo) [inline src/components/views/dialogs/AddressPickerDialog.js]: Why don't you remove this setState call? I think it is useless here, no?

File: src/components/views/rooms/RoomSettings.js [hunk 7]
```diff
@@ -584,6 +594,27 @@ module.exports = React.createClass({
         }
     },
 
+    _onExternUserSettingChange: function() {
+        const self = this;
+        const QuestionDialog = sdk.getComponent("dialogs.QuestionDialog");
+        Modal.createTrackedDialog('Reject invitation', '', QuestionDialog, {
```
Comment (giomfo) [inline src/components/views/rooms/RoomSettings.js]: What do you mean by "Reject invitation" here?

File: src/components/views/rooms/RoomSettings.js [hunk 8]
```diff
@@ -841,6 +872,19 @@ module.exports = React.createClass({
...
 
+        let externUserSection = null;
+        if (this.state.join_rule !== "public") {
+            externUserSection = (
+                <div>
+                    <label htmlFor="externOpen"> { _t('Allow the externals to join this room') } : </label>
+                    <input type="checkbox" name="externOpen" ref="externOpen"
+                           onChange={this._onExternUserSettingChange}
+                           checked={this.state.access_rules === 'unrestricted'}
+                           disabled={this.state.access_rules === 'unrestricted'} />
```
Comment (giomfo) [inline src/components/views/rooms/RoomSettings.js]: Note: only the room administrators are allowed to change this value
You should disable this checkbox when the current user is not admin

File: src/components/views/rooms/RoomSettings.js [hunk 9]
```diff
@@ -247,6 +249,14 @@ module.exports = React.createClass({
             ));
         }
 
+        if (this.state.access_rules !== originalState.access_rules) {
```
Comment (giomfo) [inline src/components/views/rooms/RoomSettings.js]: I think this block should be removed from `_calcSavePromises` function
You handle directly the new access rule  in the dialog displayed on checkbox change, no?

I mean there should be no update of the access rule rule during saving...

File: src/components/views/dialogs/AddressPickerDialog.js (hunk 6 above)
Comment (giomfo) [reply src/components/views/dialogs/AddressPickerDialog.js]: sorry I was wrong

File: src/components/views/dialogs/AddressPickerDialog.js (hunk 5 above)
Comment (giomfo) [reply src/components/views/dialogs/AddressPickerDialog.js]: ok, i was wrong
you handle here only the cases where `_lookupThreepid` returns something

File: src/components/views/dialogs/AddressPickerDialog.js (hunk 4 above)
Comment (Yoshiin) [reply src/components/views/dialogs/AddressPickerDialog.js]: I use self line 447

File: src/components/views/rooms/RoomSettings.js (hunk 9 above)
Comment (Yoshiin) [reply src/components/views/rooms/RoomSettings.js]: Right

File: src/components/views/rooms/RoomSettings.js (hunk 7 above)
Comment (Yoshiin) [reply src/components/views/rooms/RoomSettings.js]: This is only a Piwik hook string. I will change it.

File: src/components/views/rooms/RoomSettings.js (hunk 8 above)
Comment (Yoshiin) [reply src/components/views/rooms/RoomSettings.js]: I forgot this.

File: src/components/views/rooms/RoomSettings.js [hunk 10]
```diff
@@ -728,6 +723,14 @@ module.exports = React.createClass({
             currentUserLevel = defaultUserLevel;
         }
 
+        roomMembers.forEach(m => {
```
Comment (giomfo) [inline src/components/views/rooms/RoomSettings.js]: You don't need to parse all the room members, you should retrieve only the member related to the current user thanks to https://github.com/matrix-org/matrix-js-sdk/blob/85b5849228dd1bdfdf32a468559f42484f77d5c9/src/models/room.js#L880, and check his `powerLevel` (see https://github.com/matrix-org/matrix-js-sdk/blob/a0456dc43041af4528e57dfd393da701a4692732/src/models/room-member.js#L40)

Comment (Yoshiin) [reply src/components/views/rooms/RoomSettings.js]: You'r right.

File: src/components/views/rooms/RoomSettings.js [hunk 11]
```diff
@@ -143,6 +143,8 @@ module.exports = React.createClass({
             // Default to false if it's undefined, otherwise react complains about changing
             // components from uncontrolled to controlled
             isRoomPublished: this._originalIsRoomPublished || false,
+            access_rules: this._yankValueFromEvent("im.vector.room.access_rules", "rule", "restricted"),
+            externAllowed: false,
```
Comment (giomfo) [inline src/components/views/rooms/RoomSettings.js]: What is the meaning of this boolean here? isn't it redundant with the access_rule?

Comment (Yoshiin) [reply src/components/views/rooms/RoomSettings.js]: It was an old way to do the job. It can now be removed.

**Pull request merged**